### PR TITLE
Use directly the presentation compiler thread

### PR DESCRIPTION
**Pull request by mlachkar**
I think it could be nice to remove this unnecessary complexity.

### CODE REVIEW

**Review by bjaglin — COMMENTED** [review]
That would be nice indeed! Doesn't https://github.com/scalacenter/scalafix/pull/1272 showed that handling failures and restarting the compiler was sometimes necessary though (otherwise there would have been no leak?). I ran into into https://github.com/scalameta/metals/pull/2476/files#r570479181 and I have the same concerns basically.

Maybe you could smoke-test this against https://github.com/scalacenter/scalafix/issues/1229#issuecomment-680087256 and other large open source projects that use scalafix with `ExplicitResultTypes` (via a github search)? I had written instructions on how to use a scalafix-interfaces snapshot with sbt-scalafix in https://github.com/scalacenter/scalafix/issues/1299#issuecomment-745187397.

**Comment by mlachkar** [comment]
We would not have a leak if we use directly the the presentation compiler #1229, I profiled both solutions. 
In metals issue, they first considered removing the hijackPresentationCompiler, before using the same fix than in Scalafix. 
I do think that the code we have that do that was a copy from metals. 
The only change compared to presentation compiler thread is in this https://github.com/scalacenter/scalafix/blob/master/scalafix-rules/src/main/scala/scala/tools/nsc/interactive/ScalafixGlobalThread.scala#L46-L62

I agree that we need to do some testing to make sure it doesn't have any impact. I will add those tests in this Pull request

**Comment by mlachkar** [comment]
I tried the 0.9.26 and this version on Zio code base, and the profile of cpu/memory is the same: what I was expecting.


`0.9.26` : run in `[success] Total time: 86 s (01:26), completed 29 Mar 2021, 11:27:26`


image



This PR: `[success] Total time: 82 s (01:22), completed 29 Mar 2021, 11:31:26`

image

I tried both with the same conditions: I compiled first the entire project, then I run scalafixAll.
I don't think we can conclude that this PR performs better, but we can conclude that the performance should be similar. 
What do you think @bjaglin ?

File: project/Mima.scala [hunk 1]
```diff
@@ -17,7 +17,17 @@ object Mima {
...
+      ProblemFilters.exclude[MissingTypesProblem]("scala.meta.internal.pc.ScalafixGlobal"),
+      ProblemFilters.exclude[DirectMissingMethodProblem]("scala.meta.internal.pc.ScalafixGlobal.presentationCompilerThread"),
+      ProblemFilters.exclude[MissingClassProblem]("scala.tools.nsc.interactive.ScalafixGlobalProxy"),
+      ProblemFilters.exclude[DirectMissingMethodProblem]("scala.meta.internal.pc.ScalafixGlobal.hijackPresentationCompilerThread"),
+      ProblemFilters.exclude[DirectMissingMethodProblem]("scala.meta.internal.pc.ScalafixGlobal.isHijacked"),
+      ProblemFilters.exclude[DirectMissingMethodProblem]("scala.meta.internal.pc.ScalafixGlobal.metalsAsk"),
+      ProblemFilters.exclude[MissingClassProblem]("scala.tools.nsc.interactive.ScalafixGlobalThread"),
+      ProblemFilters.exclude[MissingClassProblem]("scala.tools.nsc.interactive.ScalafixGlobalThread$"),
+      ProblemFilters.exclude[MissingClassProblem]("scala.tools.nsc.interactive.InterruptException"),
+      ProblemFilters.exclude[MissingClassProblem]("scala.tools.nsc.interactive.InterruptException$")
```
Comment (bjaglin) [inline project/Mima.scala:30]: Maybe classes that are outside the scalafix namespace just to access internals should be fully excluded?
```suggestion
      ProblemFilters.exclude[MissingTypesProblem]("scala.meta.internal.*"),
      ProblemFilters.exclude[MissingClassProblem]("scala.tools.nsc.interactive.*")
```

Comment (mlachkar) [reply project/Mima.scala:30]: better, in fact.

**Pull request merged**